ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs

```diff
@@ -0,0 +1,513 @@
+    }
+}
+
+pub enum CacheEntry {
+    GzipCompressionProto(Vec<u8>),
+    // Only used for legacy cache entry.
+    Base64UncompressedProto(Vec<u8>),
+}
+
+impl CacheEntry {
+    pub fn new(bytes: Vec<u8>, storage_format: StorageFormat) -> Self {
+        match storage_format {
+            StorageFormat::GzipCompressedProto => Self::GzipCompressionProto(bytes),
+            // Legacy format.
+            StorageFormat::Base64UncompressedProto => Self::Base64UncompressedProto(bytes),
+            StorageFormat::JsonBase64UncompressedProto => {
+                panic!("JsonBase64UncompressedProto is not supported.")
+            },
+        }
+    }
+
+    pub fn into_inner(self) -> Vec<u8> {
+        match self {
+            CacheEntry::GzipCompressionProto(bytes) => bytes,
+            CacheEntry::Base64UncompressedProto(bytes) => bytes,
+        }
+    }
+
+    pub fn size(&self) -> usize {
+        match self {
+            CacheEntry::GzipCompressionProto(bytes) => bytes.len(),
+            CacheEntry::Base64UncompressedProto(bytes) => bytes.len(),
+        }
+    }
+
+    pub fn from_transaction(transaction: Transaction, storage_format: StorageFormat) -> Self {
+        let mut bytes = Vec::new();
+        transaction
+            .encode(&mut bytes)
+            .expect("proto serialization failed.");
+        match storage_format {
+            StorageFormat::GzipCompressedProto => {
+                let mut compressed = GzEncoder::new(bytes.as_slice(), flate2::Compression::fast());
+                let mut result = Vec::new();
+                compressed
+                    .read_to_end(&mut result)
+                    .expect("Gzip compression failed.");
+                CacheEntry::GzipCompressionProto(result)
+            },
+            StorageFormat::Base64UncompressedProto => {
+                let base64 = base64::encode(bytes).into_bytes();
+                CacheEntry::Base64UncompressedProto(base64)
+            },
+            StorageFormat::JsonBase64UncompressedProto => {
+                // This is fatal to see that we are using legacy file format in cache side.
+                panic!("JsonBase64UncompressedProto is not supported in cache.")
+            },
+        }
+    }
+
+    pub fn build_key(version: u64, storage_format: StorageFormat) -> String {
+        match storage_format {
+            StorageFormat::GzipCompressedProto => {
+                format!("gz:{}", version)
+            },
+            StorageFormat::Base64UncompressedProto => {
+                format!("{}", version)
+            },
+            StorageFormat::JsonBase64UncompressedProto => {
+                // This is fatal to see that we are using legacy file format in cache side.
+                panic!("JsonBase64UncompressedProto is not supported in cache.")
+            },
+        }
+    }
+}
+
+impl TryFrom<CacheEntry> for Transaction {
+    type Error = anyhow::Error;
+
+    fn try_from(value: CacheEntry) -> Result<Self, Self::Error> {
```

Comment:
could we move this into a function like `CacheEntry::to_transaction`?

Reply:
done.